### PR TITLE
elmPackages.{elm-test,elm-analyse..}: tooling additions and fixes

### PR DESCRIPTION
**Pull request by turboMaCk**
## tl;dr
This adds missing and fixes some existing tooling for elm-lang ecosystem

## Motivation
- `nodePackages.elm-test` is broken
- many commonly used tools are missing
- many tools fail to install via npm due to the binwrap issue.

This PR is based on [nix-elm-tools](https://github.com/turboMaCk/nix-elm-tools)

**detailed info can be found in commit message**

few additions remarks which can become dated:

- elmi-to-json is compiled from source using cabal2nix similar to elm-format
- there is one patch included - PR is opened in upstream package https://github.com/stoeffel/elmi-to-json/pull/28

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md). **[I hope]**

---

### CODE REVIEW

**Comment by turboMaCk** [comment]
@avh4 @BrianHicks and @domenkozar might be interested in reviewing this. Sorry for rotating_light - feel free to ignore if you're busy or not interested.

File: pkgs/development/node-packages/node-packages-v10.json [hunk 1]
```diff
@@ -21,10 +21,10 @@
 , "dhcp"
 , "dnschain"
 , "elasticdump"
-, "elm-doc-preview"
+// , "elm-doc-preview" - surpassed by elmPackages.elm-doc-preview https://github.com/NixOS/nixpkgs/pull/62899
 , "elm-live"
 , "elm-oracle"
-, "elm-test"
+// , "elm-test" - surpassed by elmPackages.elm-doc-preview https://github.com/NixOS/nixpkgs/pull/62899
```
Comment (turboMaCk) [inline pkgs/development/node-packages/node-packages-v10.json]: should be probably removed but I feel regenerating all node10 deps is out of scope of this PR.

File: pkgs/development/compilers/elm/packages/patches/elmi-to-json.patch [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
++  - bytestring
++  - containers
++  - directory
++  - filepath
++  - optparse-applicative
++  - safe-exceptions
++  - text
+   ghc-options:
+   - -Wall
+   - -Werror
```
Comment (turboMaCk) [inline pkgs/development/compilers/elm/packages/patches/elmi-to-json.patch]: PR in upstream is opened for some time but still without a response from the maintainer. https://github.com/stoeffel/elmi-to-json/pull/28

File: pkgs/development/compilers/elm/packages/generate-node-packages.sh [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env nix-shell
+#! nix-shell -i bash -p nodePackages.node2nix
+
+set -eu -o pipefail
+
+rm -f node-env.nix
+node2nix --nodejs-10 -i node-packages.json -o node-packages.nix -c node-composition.nix
```
Comment (turboMaCk) [inline pkgs/development/compilers/elm/packages/generate-node-packages.sh]: similar to https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/node-packages/generate.sh

File: pkgs/development/compilers/elm/packages/README.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+## Binwrap Patch
+
+Some node packages might use [binwrap](https://github.com/avh4/binwrap) typically for installing
+[elmi-to-json](https://github.com/stoeffel/elmi-to-json). Binwrap is not compatible with nix.
+To overcome issues with those packages apply [patch-binwrap.nix](patch-binwrap.nix) which essentially does 2 things.
+
+1. It replaces binwrap scripts with noop shell scripts
+2. It uses nix for installing the binaries to expected location in `node_modeules`
+
+Example usage be found in `elm/default.nix`.
```
Comment (turboMaCk) [inline pkgs/development/compilers/elm/packages/README.md]: please review my ~english~ czenglish

File: maintainers/maintainer-list.nix [hunk 5]
```diff
@@ -5712,4 +5712,9 @@
     github = "zzamboni";
     name = "Diego Zamboni";
   };
+  turbomack = {
+    email = "marek.faj@gmail.com";
+    github = "turboMaCk";
+    name = "Marek Fajkus";
+  };
```
Comment (turboMaCk) [inline maintainers/maintainer-list.nix]: not sure if this is appropriate given I'm not a commiter - let me know if this should be excluded.

File: pkgs/development/compilers/elm/packages/README.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+from this file. Use [generate-node-packages.sh](generate-node-packages.sh) for updates of nix expressions.
+
+## Binwrap Patch
+
+Some node packages might use [binwrap](https://github.com/avh4/binwrap) typically for installing
+[elmi-to-json](https://github.com/stoeffel/elmi-to-json). Binwrap is not compatible with nix.
+To overcome issues with those packages apply [patch-binwrap.nix](patch-binwrap.nix) which essentially does 2 things.
+
+1. It replaces binwrap scripts with noop shell scripts
+2. It uses nix for installing the binaries to expected location in `node_modeules`
```
Comment (BrianHicks) [inline pkgs/development/compilers/elm/packages/README.md]: small typo:
```suggestion
2. It uses nix for installing the binaries to expected location in `node_modules`
```

**Review by BrianHicks — COMMENTED** [review]
thank you for getting this all in one place! ❤️

I'm not a maintainer here so I can't give an approval or request CI, but since you pinged me as an Elm person it all looks reasonable. 😄

I have only ever submitted one nixpkgs change so I do not know how important this actually is, but the commit message here does not match the prescribed format: https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes

File: pkgs/development/compilers/elm/packages/README.md (hunk 6 above)
Comment (turboMaCk) [reply pkgs/development/compilers/elm/packages/README.md]: It's not just about getting this merged but about getting in merged in the best possible state. Every :eyes: on this can help it.

thanks for spotting this!

File: pkgs/development/compilers/elm/default.nix [hunk 7]
```diff
@@ -30,12 +39,25 @@ let
...
             elmVersion = elmPkgs.elm.version;
+
+            /*
+            Node/NPM based dependecies can be upgraded using script
+            `packages/generate-node-packages.sh`.
+            Packages which rely on `bin-wrap` will fail by default
+            and can be patched using `patchBinwrap` function defined in `packages/patch-binwrap.nix`.
+            */
+            elm-test = patchBinwrap [elmi-to-json] elmNodePackages.elm-test;
+            elm-verify-examples = patchBinwrap [elmi-to-json] elmNodePackages.elm-verify-examples;
```
Comment (avh4) [inline pkgs/development/compilers/elm/default.nix]: Wouldn't it be better to have elm-test and elm-verify-examples use patches on their own source code to remove the dependency on the npm elmi-to-json, instead of having them depend on a elmi-to-json with a patched binwrap?

File: maintainers/maintainer-list.nix (hunk 5 above)
Comment (domenkozar) [reply maintainers/maintainer-list.nix]: You can be a maintainer without being a committer :)

File: pkgs/development/compilers/elm/packages/patches/elmi-to-json.patch (hunk 2 above)
Comment (domenkozar) [reply pkgs/development/compilers/elm/packages/patches/elmi-to-json.patch]: you can just use `doJailbreak elmi-to-json` combinator to avoid patching.

File: pkgs/development/node-packages/node-packages-v10.json (hunk 1 above)
Comment (domenkozar) [reply pkgs/development/node-packages/node-packages-v10.json]: JSON doesn't support comments so you can just remove them and next time someone regenerates it will be gone.

**Review by domenkozar — CHANGES_REQUESTED** [review]
Good work, commented how to get some good practices in place :)

File: pkgs/development/node-packages/node-packages-v10.json (hunk 1 above)
Comment (turboMaCk) [reply pkgs/development/node-packages/node-packages-v10.json]: yes, I thought this will force people to remove it but at least it would be clear why it disappeared from generated files. Anyway, I can remove it :+1:

File: pkgs/development/compilers/elm/default.nix (hunk 7 above)
Comment (turboMaCk) [reply pkgs/development/compilers/elm/default.nix]: well, it would be possible for sure. In the end, it's a sort of tradeoff. This solution is likely to be cheaper to maintain going forward I guess.

File: pkgs/development/compilers/elm/packages/patches/elmi-to-json.patch (hunk 2 above)
Comment (turboMaCk) [reply pkgs/development/compilers/elm/packages/patches/elmi-to-json.patch]: I have an issue with `doJailbreak`. Apparently because of elmi-to-nix uses `packages.yaml` and doesn't contain `.cabal` file. I dug into `haskell.lib` using repl but it seems to me there is not build in a way to deal with it so I think patch might be still the best option.

```
jailbreak-cabal: dieVerbatim: user error (jailbreak-cabal: Error Parsing: file "elmi-to-json.cabal" doesn't exist.
```

File: pkgs/development/compilers/elm/default.nix (hunk 7 above)
Comment (turboMaCk) [reply pkgs/development/compilers/elm/default.nix]: feel free to submit suggestion as I'm not entirely sure if I understand you correctly.

File: pkgs/development/compilers/elm/packages/patches/elmi-to-json.patch (hunk 2 above)
Comment (domenkozar) [reply pkgs/development/compilers/elm/packages/patches/elmi-to-json.patch]: That's a bug in nixpkgs then, jailbreak should come after hpack has been ran.

Comment (domenkozar) [reply pkgs/development/compilers/elm/packages/patches/elmi-to-json.patch]: https://github.com/NixOS/cabal2nix/issues/421

Comment (turboMaCk) [reply pkgs/development/compilers/elm/packages/patches/elmi-to-json.patch]: this was merged in upstream so I think we now ca avoid both `doJailbreak` and patch. https://github.com/stoeffel/elmi-to-json/pull/28

Comment (turboMaCk) [reply pkgs/development/compilers/elm/packages/patches/elmi-to-json.patch]: @domenkozar elmi-to-json updated. I think this is all so this is now ready to be review again.

File: pkgs/development/compilers/elm/packages/node-packages.json [hunk 8]
```diff
@@ -0,0 +1,7 @@
+[
+    "elm-test",
+    "elm-verify-examples",
+    "elm-doc-preview",
+    { "elm-analyse": "0.16.3" },
```
Comment (turboMaCk) [inline pkgs/development/compilers/elm/packages/node-packages.json]: possible fix for 0.16.4 https://github.com/stil4m/elm-analyse/pull/208

**Comment by turboMaCk** [comment]
:thinking: not sure what happens with this

**Comment by worldofpeace** [comment]
I don't recall merging this... And it says `0` commits were merged so it looks like GitHub was maybe having a hard day :rofl:

**Comment by turboMaCk** [comment]
I think it was my screwup and potentially even GitHub issue (not really sure if an issue or on purpose behavior). Probably a rebase made gh think this is merged.
